documentation/proposals/Proposal - Region Specific Key Input.md

```diff
@@ -0,0 +1,105 @@
+# Summary
+- Make the windowing API return the key actually pressed, or any other key for that matter, instead of the key that is at
+the same location on a US-Keyboard.
+
+# Contributors
+- Matija Brown, None
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decissions
+- The main aspect is to change the original API as little as possible and also make the region-adapting
+toggleable / add the abillity to manually set a keyboard layout. To achieve this diversity,
+new layouts can be imported from an external file, but the included ones are hard coded. This way one can easilly switch between
+layouts and / or import a new, custome one while the rest of the input API remains unchanged. Due to the fact this is toggleable,
+it will also work the same way as before. The layout is saved in the IInputContext, as it is not possible to have different
+keyboard layouts on different keyboards at the same time, so it needn't be saved in every keyboard instance. If writing custom layouts
+turns out to be a common thing, there will be a layout generator created as well. For now key modifiers such as SHIFT will not be
+considered, but that might be implemented at a later time.
+
+# Proposed API
+
+## KeyboardLayout
+```cs
+public class KeyboardLayout
+{
+
+    /// <summary>
+    /// The different included layouts.
+    /// </summary>
+    public enum Layouts
```

Comment:
I don't think an enum is the right choice here - I feel using inheritance for this would be better.

Reply:
One could definitely use inheritance there, the reason I put an enumeration is because this is only meant for the user to set the keyboard layout. I thought about each of the layouts being abstracted from the KeyboardLayout class, but that would mean having to write an own implementation for custom keyboards, while this way one can easily just set a flag. But I guess inheritance would be possible as well, if it fits better.

Reply:
My idea would've been that we have some pre-defined instances like `KeyboardLayout.QWERTY` which would just implement an interface. I think that'd make for the most flexibility while keeping usage equally simple. Of course some boilerplate on our side, but I don't think it's too much

Reply:
That's probably the best way, I agree, but what about custom layouts? Just have a ```KeyboardLayout.Custom```? That would work, just be a lot different from the others.

Reply:
That'd be a problem with enums, but with static properties that just return different implementations of an interface, a custom layout would just not work via `KeyboardLayout`
instead I'd be something like `Layout = new MyCustomKeyboardLayout()`

Reply:
I mean, I could have a method ```KeyboardLayout.getCusomLayout("myCustomKeyboardLayout.slkLyt");```. Or Just a custom one, but I guess one could also do ```KeyboardLayout.Custom.Create("file.slkLyt");```? The problem I see with the custom class is that one would have to write an own file parser, although that could be supplied as well, if it were done that way.

Reply:
I'm guessing `slkLyt` is SilkLayout? So not some standard format?
Why is that favourable over
```cs
public sealed class MyCustomLayout : IKeyboardLayout
{
    public Key MapKey(Key original)
    {
        // probably a switch statement or whatever
    }
}
```

I guess having those files mean the user of our user can modify it, but I'm not sure if that's a thing we want to force?

Reply:
That is just a way of storing the layout. (I just made up a file extension, in this case yes, silkLayout). It just maps a key code to another, just like that method. Only, instead of the giant switch statement, I thought one could use a Dictionary to map the keys and load the data from a file. That would mean not writing these long switch statements and easier configurabillity. Also, then it would be possible to have a generic ```Customlayout``` class that just takes in the file.

Reply:
I mean I guess we can provide a separate type that loads a file - but I don't think the inbuilt ones should also be based on such files.

Reply:
Sounds good. So I guess I'll change the proposal documentation to be up to date, and then contact you again.

Reply:
I have committed a few changes to the proposal file. Is this better?